File: client_example.c
/*
 * Example client for Switch Controller API
 * Demonstrates how to control the virtual controller
 *
 * Compile: gcc -o client_example client_example.c
 * Run: ./client_example
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/un.h>

#define SOCKET_PATH "/tmp/switch_controller.sock"

/* Send command to controller */
static int send_command(const char *cmd) {
    int sock;
    struct sockaddr_un addr;
    char response[256];

    sock = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sock < 0) {
        perror("socket");
        return -1;
    }

    memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    strncpy(addr.sun_path, SOCKET_PATH, sizeof(addr.sun_path) - 1);

    if (connect(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        perror("connect");
        close(sock);
        return -1;
    }

    if (write(sock, cmd, strlen(cmd)) < 0) {
        perror("write");
        close(sock);
        return -1;
    }

    ssize_t n = read(sock, response, sizeof(response) - 1);
    if (n > 0) {
        response[n] = '\0';
        printf("Response: %s", response);
    }

    close(sock);
    return 0;
}

/* Button press helper */
static void press_button(const char *name, unsigned int mask) {
    char cmd[64];
    printf("Pressing %s button...\n", name);
    snprintf(cmd, sizeof(cmd), "BTN %x", mask);
    send_command(cmd);
    usleep(100000); /* Hold for 100ms */
    send_command("BTN 0"); /* Release */
    usleep(200000); /* Wait 200ms */
}

/* Move stick helper */
static void move_stick(const char *direction) {
    printf("Moving left stick %s...\n", direction);

    if (strcmp(direction, "up") == 0) {
        send_command("STICK 128 0 128 128");
    } else if (strcmp(direction, "down") == 0) {
        send_command("STICK 128 255 128 128");
    } else if (strcmp(direction, "left") == 0) {
        send_command("STICK 0 128 128 128");
    } else if (strcmp(direction, "right") == 0) {
        send_command("STICK 255 128 128 128");
    }

    usleep(500000); /* Hold for 500ms */
    send_command("STICK 128 128 128 128"); /* Center */
    usleep(200000);
}

/* D-pad helper */
static void press_dpad(const char *direction) {
    char cmd[64];
    int hat;

    printf("Pressing D-pad %s...\n", direction);

    if (strcmp(direction, "up") == 0) hat = 0;
    else if (strcmp(direction, "right") == 0) hat = 2;
    else if (strcmp(direction, "down") == 0) hat = 4;
    else if (strcmp(direction, "left") == 0) hat = 6;
    else return;

    snprintf(cmd, sizeof(cmd), "HAT %d", hat);
    send_command(cmd);
    usleep(100000);
    send_command("HAT 8"); /* Neutral */
    usleep(200000);
}

int main(int argc, char **argv) {
    printf("Switch Controller Client Example\n");
    printf("=================================\n\n");

    /* Reset to neutral state */
    printf("Resetting controller...\n");
    send_command("RESET");
    usleep(500000);

    /* Press A button (confirm) */
    press_button("A", 0x0004); /* BTN_A */

    /* Press B button (back) */
    press_button("B", 0x0002); /* BTN_B */

    /* Press X and Y together */
    printf("Pressing X+Y buttons...\n");
    send_command("BTN 0x0009"); /* BTN_Y | BTN_X */
    usleep(100000);
    send_command("BTN 0");
    usleep(200000);

    /* Move left stick in a circle */
    printf("Moving left stick in a circle...\n");
    move_stick("up");
    move_stick("right");
    move_stick("down");
    move_stick("left");

    /* Use D-pad */
    press_dpad("up");
    press_dpad("down");
    press_dpad("left");
    press_dpad("right");

    /* Press + button (menu) */
    press_button("Plus", 0x0200); /* BTN_PLUS */

    /* Press Home button */
    press_button("Home", 0x1000); /* BTN_HOME */

    printf("\nDemo complete!\n");
    return 0;
}

File: client_example.py
#!/usr/bin/env python3
"""
Python client example for Switch Controller API
Demonstrates how to control the virtual controller from Python
"""

import socket
import time

SOCKET_PATH = "/tmp/switch_controller.sock"

# Button masks
BTN_Y = 0x0001
BTN_B = 0x0002
BTN_A = 0x0004
BTN_X = 0x0008
BTN_L = 0x0010
BTN_R = 0x0020
BTN_ZL = 0x0040
BTN_ZR = 0x0080
BTN_MINUS = 0x0100
BTN_PLUS = 0x0200
BTN_LCLICK = 0x0400
BTN_RCLICK = 0x0800
BTN_HOME = 0x1000
BTN_CAPTURE = 0x2000

# HAT positions
HAT_UP = 0
HAT_UP_RIGHT = 1
HAT_RIGHT = 2
HAT_DOWN_RIGHT = 3
HAT_DOWN = 4
HAT_DOWN_LEFT = 5
HAT_LEFT = 6
HAT_UP_LEFT = 7
HAT_NEUTRAL = 8


def send_command(cmd):
    """Send a command to the controller API"""
    try:
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(SOCKET_PATH)
        sock.send(cmd.encode())
        response = sock.recv(256).decode()
        sock.close()
        return response.strip()
    except Exception as e:
        print(f"Error: {e}")
        return None


def press_button(name, mask, duration=0.1):
    """Press and release a button"""
    print(f"Pressing {name} button...")
    send_command(f"BTN {mask:04x}")
    time.sleep(duration)
    send_command("BTN 0")
    time.sleep(0.2)


def move_stick(stick, x, y, duration=0.5):
    """Move a stick to a position
    stick: 'left' or 'right'
    x, y: 0-255 (128 is center)
    """
    if stick == 'left':
        send_command(f"STICK {x} {y} 128 128")
    else:
        send_command(f"STICK 128 128 {x} {y}")
    time.sleep(duration)
    send_command("STICK 128 128 128 128")  # Center
    time.sleep(0.2)


def press_dpad(direction):
    """Press a D-pad direction"""
    directions = {
        'up': HAT_UP,
        'right': HAT_RIGHT,
        'down': HAT_DOWN,
        'left': HAT_LEFT
    }
    if direction in directions:
        print(f"Pressing D-pad {direction}...")
        send_command(f"HAT {directions[direction]}")
        time.sleep(0.1)
        send_command(f"HAT {HAT_NEUTRAL}")
        time.sleep(0.2)


def reset():
    """Reset controller to neutral state"""
    print("Resetting controller...")
    send_command("RESET")
    time.sleep(0.2)


def main():
    print("Switch Controller Python Client Example")
    print("=" * 40)
    print()

    # Reset to neutral state
    reset()

    # Press individual buttons
    press_button("A", BTN_A)
    press_button("B", BTN_B)
    press_button("X", BTN_X)
    press_button("Y", BTN_Y)

    # Press multiple buttons together
    print("Pressing L+R together...")
    send_command(f"BTN {BTN_L | BTN_R:04x}")
    time.sleep(0.1)
    send_command("BTN 0")
    time.sleep(0.5)

    # Move left stick in a circle
    print("Moving left stick in a circle...")
    move_stick('left', 128, 0)    # Up
    move_stick('left', 255, 0)    # Up-right
    move_stick('left', 255, 128)  # Right
    move_stick('left', 255, 255)  # Down-right
    move_stick('left', 128, 255)  # Down
    move_stick('left', 0, 255)    # Down-left
    move_stick('left', 0, 128)    # Left
    move_stick('left', 0, 0)      # Up-left

    # Use D-pad
    press_dpad('up')
    press_dpad('down')
    press_dpad('left')
    press_dpad('right')

    # Press special buttons
    press_button("Plus", BTN_PLUS)
    press_button("Minus", BTN_MINUS)

    # Combo: Take a screenshot (Capture button)
    press_button("Capture", BTN_CAPTURE, duration=0.1)

    # Combo: Return to home screen
    press_button("Home", BTN_HOME)

    print()
    print("Demo complete!")


if __name__ == "__main__":
    main()
